src/server/builder.rs

```diff
@@ -0,0 +1,181 @@
+use actix_cors::Cors;
+use actix_web::{
+    App, HttpServer,
+    middleware::Logger,
+};
+use dotenv::dotenv;
+use log::{info, error};
+
+#[cfg(not(feature = "swagger"))]
+use actix_web::web;
+
+#[cfg(feature = "swagger")]
+use paperclip::{
+    actix::{web, OpenApiExt},
+    v2::models::DefaultApiRaw,
+};
+
+use crate::server::default_cors;
+
+use super::threads;
+
+use super::{
+    stats::{BaseStats, StatsWrapper, StatsPresenter, AppDataWrapper, default_healthcheck_handler, default_readiness_handler, default_stats_handler},
+    logger,
+};
+
+pub struct Microservice<'a> {
+    app_port: &'a str,
+    run_env: &'a str,
+    #[cfg(feature = "swagger")]
+    swagger_mount: &'a str,
+    #[cfg(feature = "swagger")]
+    swagger_spec: DefaultApiRaw,
+}
+
+impl Default for Microservice<'_>
+{
+    fn default() -> Self {
+        Microservice {
+            app_port: "8000",
+            run_env: "dev",
+            #[cfg(feature = "swagger")]
+            swagger_mount: "/swagger",
+            #[cfg(feature = "swagger")]
+            swagger_spec: DefaultApiRaw::default(),
+        }
+    }
+}
+
+impl<'a> Microservice<'a>
+{
+    pub fn set_app_port(mut self, app_port: &'a str) -> Self
+    {
+        self.app_port = app_port;
+        self
+    }
+
+    pub fn set_run_env(mut self, run_env: &'a str) -> Self
+    {
+        self.run_env = run_env;
+        self
+    }
+
+    #[cfg(feature = "swagger")]
+    pub fn set_swagger_mount(mut self, swagger_mount: &'a str) -> Self
+    {
+        self.swagger_mount = swagger_mount;
+        self
+    }
+
+    #[cfg(feature = "swagger")]
+    pub fn set_swagger_spec(mut self, swagger_spec: DefaultApiRaw) -> Self
+    {
+        self.swagger_spec = swagger_spec;
+        self
+    }
+
+    #[cfg(feature = "swagger")]
+    pub fn set_swagger_info(mut self, pkg_name: impl Into<String>, pkg_version: impl Into<String>, pkg_description: impl Into<String>) -> Self
+    {
+        self.swagger_spec.info.title = pkg_name.into();
+        self.swagger_spec.info.version = pkg_version.into();
+        self.swagger_spec.info.description = Some(pkg_description.into());
+        self
+    }
+
+    pub async fn start<D, T, F, C> (
+        self,
+        prepare_app_data: impl Fn() -> T + Sized,
+        configure_app: F,
+        cors_factory: Option<C>,
+    ) -> std::io::Result<()>
+    where
+        D: AppDataWrapper + 'static,
+        T: StatsPresenter<D> + 'static + Clone + Send + Sync,
+        F: Fn(&mut web::ServiceConfig) + Send + Clone + 'static + Sized,
+        C: Fn() -> Cors + Send + Clone + 'static,
+    {
+        if let Some(cors_factory) = cors_factory {
+            self.start_inner(prepare_app_data, configure_app, cors_factory).await
+        } else {
+            self.start_inner(prepare_app_data, configure_app, default_cors).await
+        }
```

Comment:
consider extracting cors config to variable, such that:
```suggestion
        let cors = cors_factory.unwrap_or(default_cors);
        self.start_inner(prepare_app_data, configure_app, cors).await
```

Reply:
Can't do it, cors_factory is of type `C: Fn() -> Cors + Send + Clone + 'static` while default_cors is of type `fn() -> actix_cors::Cors`